statsmodels/gam/tests/test_gam.py

```diff
@@ -0,0 +1,791 @@
+
+    for _ in range(10):
+        params = np.random.randint(-2, 2, 5)
+        gam_der2 = gp.deriv2(params)
+        hess = hessian(params)
+        hess = np.flipud(hess)
+        hess = np.fliplr(hess)
+        assert_allclose(gam_der2, hess, atol=1.e-13, rtol=1.e-3)
+
+
+def test_approximation():
+    np.random.seed(1)
+    poly, y = polynomial_sample_data()
+    alpha = 1
+    for _ in range(10):
+        params = np.random.uniform(-1, 1, 4)
+        cost, err, itg = cost_function(params, poly, y, alpha)
+        glm_gam = GLMGam(y, smoother=poly, alpha=alpha)
+        # TODO: why do we need pen_weight=1
+        gam_loglike = glm_gam.loglike(params, scale=1, pen_weight=1)
+        assert_allclose(err - itg, cost, rtol=1e-10)
+        assert_allclose(gam_loglike, cost, rtol=0.1)
+
+
+def test_gam_glm():
+    cur_dir = os.path.dirname(os.path.abspath(__file__))
+    file_path = os.path.join(cur_dir, "results", "prediction_from_mgcv.csv")
+    data_from_r = pd.read_csv(file_path)
+    # dataset used to train the R model
+    x = data_from_r.x.as_matrix()
+    y = data_from_r.y.as_matrix()
+
+    df = [10]
+    degree = [3]
+    bsplines = BSplines(x, degree=degree, df=df, include_intercept=True)
+    # y_mgcv is obtained from R with the following code
+    # g = gam(y~s(x, k = 10, bs = "cr"), data = data, scale = 80)
+    y_mgcv = np.asarray(data_from_r.y_est)
+
+    # alpha = 1000
+    alpha = 0.03
+    alpha = 0.0001
+    alpha = 0.1
+
+    glm_gam = GLMGam(y, smoother=bsplines, alpha=alpha)
+    res_glm_gam = glm_gam.fit(method='bfgs', max_start_irls=0,
+                              disp=1, maxiter=10000, maxfun=5000)
+
+    glm_gam = GLMGam(y, smoother=bsplines, alpha=alpha)
+
+    res_glm_gam = glm_gam.fit(method='bfgs', max_start_irls=0,
+                              disp=1, maxiter=10000, maxfun=5000)
+    y_gam0 = np.dot(bsplines.basis, res_glm_gam.params)
+    y_gam = np.asarray(res_glm_gam.fittedvalues)
+    assert_allclose(y_gam, y_gam0, rtol=1e-10)
+
+    # plt.plot(x, y_gam, '.', label='gam')
+    # plt.plot(x, y_mgcv, '.', label='mgcv')
+    # plt.plot(x, y, '.', label='y')
+    # plt.legend()
+    # plt.show()
+
+    assert_allclose(y_gam, y_mgcv, atol=1.e-2)
+
+
+def test_gam_discrete():
+    cur_dir = os.path.dirname(os.path.abspath(__file__))
+    file_path = os.path.join(cur_dir, "results", "prediction_from_mgcv.csv")
+    data_from_r = pd.read_csv(file_path)
+    # dataset used to train the R model
+    x = data_from_r.x.as_matrix()
+    y = data_from_r.ybin.as_matrix()
+
+    df = [10]
+    degree = [5]
+    bsplines = BSplines(x, degree=degree, df=df, include_intercept=True)
+
+    # y_mgcv is obtained from R with the following code
+    # g = gam(y~s(x, k = 10, bs = "cr"), data = data, scale = 80)
+    y_mgcv = data_from_r.ybin_est
+
+    alpha = 0.00002
+    # gp = UnivariateGamPenalty(alpha=alpha, univariate_smoother=bsplines)
+    # lg_gam = LogitGam(y, bsplines.basis, penal=gp)
+    #
```

Comment:
?

Reply:
TODO: this needs another look comparing LogitGAM and GLMGAM